### ISSUE TITLE
Error when executing the runtime with nvdla_small

### ISSUE THREAD

**Issue by giusecesa4** [nvdla/sw#49]
Hi everyone,
I have a problem when running tests on the nvdla small.
I put the architecture on an FPGA with the Linux OS running on it, and when I try executing the following instruction to run a test:

>  ./runtime/nvdla_runtime/nvdla_runtime --loadable tests/CDP_L0_0_small_fbuf

I have this error:

> libnvdla<1> failed to open dla device
> libnvdla<1> Out of bounds DLA instance 0 requested.
> (DLA_TEST) Error 0x00000004: runtime->load failed (in RuntimeTest.cpp, function loadLoadable(), line 265)
> (DLA_TEST) Error 0x00000004: (propagating from RuntimeTest.cpp, function run(), line 318)
> (DLA_TEST) Error 0x00000004: (propagating from main.cpp, function launchTest(), line 87)

I tried to insert the module opendla.ko and I succeded, but I am not able to generate the other module drm.ko and to insert it.
Maybe this is the problem? Does anyone have some ideas?
Thanks

**Comment by prasshantg** [comment]
@giusecesa4 please share log when you insert modules

**Comment by giusecesa4** [comment]
> [  273.282316] opendla: loading out-of-tree module taints kernel.

This is the only thing I have

**Comment by giusecesa4** [comment]
When I try to insert the drm module, instead, I have this error:

```
[   28.920957] drm: Unknown symbol __check_object_size (err 0)
[   28.926530] drm: Unknown symbol trace_seq_printf (err 0)
[   28.931784] drm: Unknown symbol trace_raw_output_prep (err 0)
[   28.937495] drm: Unknown symbol __stack_chk_guard (err 0)
[   28.942932] drm: Unknown symbol trace_event_raw_init (err 0)
[   28.948545] drm: Unknown symbol event_triggers_call (err 0)
[   28.954094] drm: Unknown symbol kmem_cache_alloc_trace (err 0)
[   28.959896] drm: Unknown symbol trace_event_buffer_commit (err 0)
[   28.965978] drm: Unknown symbol __stack_chk_fail (err 0)
[   28.971270] drm: Unknown symbol trace_event_ignore_this_pid (err 0)
[   28.977559] drm: Unknown symbol perf_trace_buf_alloc (err 0)
[   28.983151] drm: Unknown symbol perf_trace_run_bpf_submit (err 0)
[   28.989264] drm: Unknown symbol vga_client_register (err 0)
[   28.994804] drm: Unknown symbol trace_event_reg (err 0)
[   29.000041] drm: Unknown symbol trace_define_field (err 0)
[   29.005469] drm: Unknown symbol kgdb_active (err 0)
[   29.010363] drm: Unknown symbol trace_event_buffer_reserve (err 0)
[   29.016516] drm: Unknown symbol __dynamic_pr_debug (err 0)
[   29.021988] drm: Unknown symbol from_kuid_munged (err 0)
[   29.027266] drm: Unknown symbol trace_handle_return (err 0)
[   29.032813] drm: Unknown symbol _mcount (err 0)
insmod: can't insert 'drm.ko': unknown symbol in module, or unknown parameter
```

**Comment by zx4e71** [comment]
We also try to run small on FPGA, after module load we see "reset engine done" message - which is OK and you seem to be missing that one. But nothing from DRM. Runtime then hangs with memory error. DRM device registration looks fine, but there must be something missing: on VP we see one more message: "[drm] Initialized nvdla 0.0.0 20171017 for 10200000.nvdla on minor 0" which does not appear on FPGA. Any suggestions why drm device seems to be correctly initialized but still it fails?

**Comment by prasshantg** [comment]
@giusecesa4 looks some mismatch between kernel and drm module you are using. Please share full steps that you followed.

@zx4e71 it is possible that device tree for FPGA does not contain the required node and hence driver probe is not getting called for FPGA

**Comment by prasshantg** [comment]
@zx4e71 @giusecesa4 any update?

**Comment by giusecesa4** [comment]
I have problems with the Runtime in the sense that I get a segmentation fault given by a PC/SP misalignment. Maybe the module opendla.ko is not inserted correctly. I am trying to run the small NVDLA on an FPGA, but for now I am stucked at this point

**Comment by smsharif1991** [comment]
Hi @giusecesa4 ,
Can you please try to use drm.ko and opendla_small.ko from prebuilts first to see if all the setup is fine?
https://github.com/nvdla/sw/tree/master/prebuilt/linux

Also make sure that you are using nv_small VP and not the default VP

Also please mention all the steps you are following.

Thanks.

**Comment by giusecesa4** [comment]
No, I can't. I am using another version of the linux kernel (4.14) and I'm not using VP but I am working on FPGA. I need to recompile the module from the sources

**Comment by zx4e71** [comment]
We were able to send single job to SDP, it seemed to do some operations, but we were not notified about job being finished with the interrupt. We turned to bare metal and we are trying to get to some conclussions on that...

**Comment by JunningWu** [comment]
@giusecesa4 when you build kmd for small, do you just add CFLAG_MODULE='-DDLA_SMALL_CONFIG' to the default commond like this:

```
make KDIR=<path_to_Linux_source> ARCH=arm64 CROSS_COMPILE=<path_to_toolchain> CFLAG_MODULE='-DDLA_SMALL_CONFIG'
```

and the generated file opendla.ko is just what we want? which mean this is opendla_small.ko?

and where do you get the newly compiled drm.ko file?

**Comment by smsharif1991** [comment]
@JunningWu  yes the newly generated opendla.ko is opendla_small.ko
You can use the same drm.ko for nv_small .

Thanks.

**Comment by anakin1028** [comment]
@zx4e71 do you have any guidance for including the required node to the device tree? I try to change from `AARCH64` to `arm-versatile` in the Virtual Platform, but the KMD is not being probed as @prasshantg  has said.

**Comment by gitosu67** [comment]
@prasshantg @anakin1028 @zx4e71 I am also getting the same error. I'm not using VP but I am trying to implement it on FPGA. I am not sure how to generate drm.ko from scratch.

**Comment by cfkibby** [comment]
I also get same issue. but no idea. @giusecesa4 do u fix this issue finally?

**Comment by gitosu67** [comment]
@cfkibby I was able to solve the issue. In nvdla_gem.c, change `if(!dma) {}` to `if(dma){}` in line 441. If you are using linux kernel 4.14+ the output of `dma_declare_coherent_memory()` has been altered.

**Comment by cfkibby** [comment]
@Okaymaddy Thank you so much for your very quick reply and hint. Ye, it is. I also read other tickets and actually update 3 places for this issue.1st one is the  if(dma) issue, 2nd one is :I update compatible attribute in device tree, it should be "nvidia,nv_small". 3rd one is kmd/firmware/include/opendla.h, add #define DLA_2_CONFIG.
Now, it seems better. when I execute:
insmod opendla_small.ko
opendla: loading out-of-tree module taints kernel.
Probe NVDLA config nvidia,nv_small
0 . 12 . 5
reset engine done
[drm] Initialized nvdla 0.0.0 20171017 for 43c00000.nvdla_v1_0 on minor 0
NVDLA 43c00000.nvdla_v1_0: failed to register drm device
NVDLA: probe of 43c00000.nvdla_v1_0 failed with error -12 

it seems that something with my hardware, I think I need to review my hardware.
Thanks.

**Comment by gitosu67** [comment]
@cfkibby I remember getting that same issue, and it turned out that I forgot the interrupt connection in my hardware design. Not sure if that is the same in your case.

**Comment by cfkibby** [comment]
@Okaymaddy Thank u for your hint, I finally solve the driver install issue.  in my environment, I just failed to reserved memory for the NVDLA, so it failed(I have no idea why updating device tree is not working). I use below method to reserve memory for nvdla,
 in arch/arm/mach-zynq/common.c function zynq_memory_init, I add below func:
memblock_free(0x30000000, 256 * SZ_1M);
memblock_remove(0x30000000, 256 * SZ_1M);
printk(“reserve 256MB@0x30000000 forPL\n”);

so that I can reserve higher 256MB for nvdla.

**Comment by shaumik1** [comment]
> > [  273.282316] opendla: loading out-of-tree module taints kernel.
> 
> This is the only thing I have

@cfkibby @giusecesa4  or anyone else: could you resolve this issue? I have tried to ensure all the points mentioned in the thread, but still get the same message.
Further, when I try `ls /dev/dri/` I do not see `renderD128` driver..

I am using Petalinux2019.2
Any recommendations?

**Comment by sunny-yellow** [comment]
> > > [  273.282316] opendla: loading out-of-tree module taints kernel.
> > 
> > 
> > This is the only thing I have
> 
> @cfkibby @giusecesa4 or anyone else: could you resolve this issue? I have tried to ensure all the points mentioned in the thread, but still get the same message.
> Further, when I try `ls /dev/dri/` I do not see `renderD128` driver..
> 
> I am using Petalinux2019.2
> Any recommendations?

Hi,shaumik. Did you finish the work for implementation NVDLA on FPGA. I hope to communicate with you for some questions.

**Comment by sunny-yellow** [comment]
> > > > [  273.282316] opendla: loading out-of-tree module taints kernel.
> > > 
> > > 
> > > This is the only thing I have
> > 
> > 
> > @cfkibby @giusecesa4 or anyone else: could you resolve this issue? I have tried to ensure all the points mentioned in the thread, but still get the same message.
> > Further, when I try `ls /dev/dri/` I do not see `renderD128` driver..
> > I am using Petalinux2019.2
> > Any recommendations?
> 
> Hi,shaumik. Did you finish the work for implementation NVDLA on FPGA. I hope to communicate with you for some questions.

I am also use Petalinux2019.2.
